ruby/phase-1.md

```diff
@@ -0,0 +1,398 @@
+managing namespaces. As hinted in the previous section there's a global singleton client available
+for applications that only connect to a single Temporal server and don't need varying
+configurations:
+
+```ruby
+# Global instance of the Client configured via `Temporal.configure`
+Temporal.start_workflow(MyWorkflow)
+
+# Explicitly configured local instance of the Client
+config = Temporal::Configuration.new(url: 'localhost:7233')
+client = Temporal::Client.new(config)
+client.start_workflow(MyWorkflow)
+```
+
+Despite the differences both clients offer exactly the same interface (the singleton client is an
+instance of a regular client stored on Temporal module for convenience). The interface will look
+something like this:
+
+```ruby
+class Client
+  def initialize(config)
+
+  def start_workflow(
+    workflow, # workflow class or a name
+    *args, # any args to be passed as the input to the workflow
+    options: {}, # execution options overrides
+    **kwargs # keyword arguments as inputs are also supported
+  ) -> WorkflowHandle
+
+  def workflow_handle(namespace, workflow_id, run_id = nil) -> WorkflowHandle
+end
+```
+
+Notes:
+
+- This is not a complete definition of the client
+- Both named and keyword arguments are supported, except for the reserved `options:`
+
+The combination of the singleton client and implicit execution attributes allows us to optionally
+support direct workflow invocations (removing all SDK traces out of the way):
+
+```ruby
+MyWorkflow.start(arg_1, arg_2, keyword_arg: arg_3)
+```
+
+Once a workflow is started all the interactions will be going through the handle object returned to
+the caller. The handle can of course be created from a namespace/workflow_id/run_id combination
+using the `Client#workflow_handle` method. Here the interface of the handle:
+
+```ruby
+class WorkflowHandle
+  def initialize(connection, namespace, workflow_id, run_id = nil)
+
+  def result(await: false, timeout: nil) -> Any
+
+  def describe -> WorkflowExecutionInfo
+
+  def cancel -> void
+
+  def query(name, *args) -> Any
+
+  def signal(name, *args) -> void
+
+  def terminate(reason, details: nil) -> void
+end
+
+```
+
+
+## Worker
+
+The worker interface allows SDK users to register any number of both workflows and activities. These
+will then be polled for on the specified namespaces & task queues and executed. The worker can poll
+on multiple task queues at the same time. These will be determined by the registered activities and
+workflow. If there are any overlaps we'll only create one poller for each unique task queue.
+
+```ruby
+config Temporal::Configuration.new(url: 'localhost:7233')
+worker = Temporal::Worker.new(config)
+
+worker.register_workflow(NewWorkflow)
+worker.register_workflow(NewWorkflow, task_queue: 'tq-3')
+worker.register_activity(AnotherActivity)
+worker.register_activity(AnotherActivity, namespace: 'ns-3')
+
+# Calling Worker#run will block the main thread until receiving SIGTERM or SIGINT
+worker.run
```

Comment:
Can we have a `worker.run(daemon: true)` call? 

Then couple it with a `worker.join` or `worker.wait` which could block on exit. Just thinking of ways to make it easier to support workers running in their own thread but a shared process with other things.

Reply:
Ah, yes, that should be possible 👍